linera-faucet/server/src/lib.rs

```diff
@@ -182,72 +214,282 @@ where
+                    // Rate limit exceeded - reject all requests
+                    for request in valid_requests {
+                        let _ = request.responder.send(Err(Error::new(
+                            "Not enough unlocked balance; try again later.",
+                        )));
+                    }
+                    return Ok(());
                 }
             }
         }
 
-        let result = client
-            .open_chain(
-                ChainOwnership::single(owner),
-                ApplicationPermissions::default(),
-                self.amount,
-            )
-            .await;
+        // Create OpenChain operations for all valid requests
+        let mut operations = Vec::new();
+        for request in &valid_requests {
+            let config = OpenChainConfig {
+                ownership: ChainOwnership::single(request.owner),
+                balance: self.config.amount,
+                application_permissions: ApplicationPermissions::default(),
+            };
+            operations.push(Operation::system(SystemOperation::OpenChain(config)));
+        }
+
+        // Execute all operations in a single block
+        let result = client.execute_operations(operations, vec![]).await;
         self.context.lock().await.update_wallet(&client).await?;
-        let description = match result? {
-            ClientOutcome::Committed((description, _certificate)) => description,
-            ClientOutcome::WaitForTimeout(timeout) => {
-                return Err(Error::new(format!(
+
+        let certificate = match result {
+            Err(ChainClientError::LocalNodeError(LocalNodeError::WorkerError(
+                WorkerError::ChainError(chain_err),
+            ))) => {
+                match *chain_err {
+                    ChainError::ExecutionError(exec_err, ChainExecutionContext::Operation(i))
+                        if i > 0
+                            && matches!(
+                                *exec_err,
+                                ExecutionError::BlockTooLarge
+                                    | ExecutionError::FeesExceedFunding { .. }
+                                    | ExecutionError::InsufficientBalance { .. }
+                                    | ExecutionError::MaximumFuelExceeded(_)
+                            ) =>
+                    {
+                        tracing::error!(%exec_err, "Execution of operation {i} failed; reducing batch size");
+                        self.config.max_batch_size = i as usize;
+                        // Put the valid requests back into the queue.
+                        let mut pending_requests = self.pending_requests.lock().await;
+                        for request in valid_requests.into_iter().rev() {
+                            pending_requests.push_front(request);
+                        }
+                        return Ok(()); // Don't return an error, so we retry.
```

Comment:
Is it OK to call `update_wallet` (we do that earlier) for the case when we end up retrying? if we didn't propose and there's no timeout, there's no need to update it.

Reply:
First of all there's no harm in doing so, apart from a bit of additional disk I/O.

But more importantly, if we made a proposal to the validators, we should _always_ save the wallet, even if the proposal failed. We might have a pending block now that we need to remember.